src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.cs

```diff
@@ -66,92 +65,170 @@ public async Task FindReferencesAsync(ISymbol symbol, CancellationToken cancella
+            ImmutableArray<ISymbol> symbols, CancellationToken cancellationToken)
+        {
+            var projects = _documents != null
+                ? _documents.Select(d => d.Project).ToImmutableHashSet()
+                : _solution.Projects.ToImmutableHashSet();
 
-                foreach (var (project, documentMap) in projectToDocumentMap)
-                    tasks.Add(Task.Factory.StartNew(() => ProcessProjectAsync(project, documentMap, cancellationToken), cancellationToken, TaskCreationOptions.None, _scheduler).Unwrap());
+            var result = new HashSet<ProjectId>();
 
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+            foreach (var symbol in symbols)
+            {
+                var dependentProjects = await DependentProjectsFinder.GetDependentProjectsAsync(
+                    _solution, symbol, projects, cancellationToken).ConfigureAwait(false);
+                foreach (var project in dependentProjects)
+                    result.Add(project.Id);
             }
+
+            return result;
         }
 
-        [Conditional("DEBUG")]
-        private static void ValidateProjectToDocumentMap(
-            ProjectToDocumentMap projectToDocumentMap)
+        private async Task ProcessProjectAsync(Project project, ImmutableArray<ISymbol> allSymbols, CancellationToken cancellationToken)
         {
-            var set = new HashSet<ISymbol>();
-
-            foreach (var documentMap in projectToDocumentMap.Values)
+            try
             {
-                foreach (var documentToFinderList in documentMap)
+                using var _1 = PooledHashSet<Document>.GetInstance(out var allDocuments);
+                foreach (var symbol in allSymbols)
                 {
-                    set.Clear();
-
-                    foreach (var tuple in documentToFinderList.Value)
-                        Debug.Assert(set.Add(tuple));
+                    foreach (var finder in _finders)
+                    {
+                        var documents = await finder.DetermineDocumentsToSearchAsync(
+                            symbol, project, _documents, _options, cancellationToken).ConfigureAwait(false);
+                        allDocuments.AddRange(documents);
+                    }
                 }
+
+                using var _2 = ArrayBuilder<Task>.GetInstance(out var tasks);
+                foreach (var document in allDocuments)
+                    tasks.Add(CreateWorkAsync(() => ProcessDocumentsAsync(document, allSymbols, cancellationToken), cancellationToken));
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+            }
+            finally
+            {
+                await _progressTracker.ItemCompletedAsync(cancellationToken).ConfigureAwait(false);
             }
         }
 
-        private async ValueTask HandleLocationAsync(ISymbol symbol, ReferenceLocation location, CancellationToken cancellationToken)
+        private async Task ProcessDocumentsAsync(
```

Comment:
ProcessDocumentSymbols perhaps?  Only processing 1 document

Reply:
Oh gosh. Yes, this is misnamed.